activerecord/lib/active_record/reflection.rb

```diff
@@ -809,6 +809,14 @@ def join_foreign_type
         foreign_type
       end
 
+      def check_validity!
+        super
+
+        if !polymorphic? && klass.composite_primary_key? && Array(association_primary_key).length != Array(foreign_key).length
```

Comment:
The validation is true for any type of the model so `klass.composite_primary_key?` check is not necessary from the design point of view. However we can kind of use it to prevent unnecessary arrays allocation for non-cpk models. So it's up to you too keep it
```suggestion
        if !polymorphic? && Array(association_primary_key).length != Array(foreign_key).length
```

If we are going to keep it I think we designed `association_primary_key` and `foreign_key` to be an array if the `klass` is a cpk model, so potentially we should be able to do:

```suggestion
        if !polymorphic? && klass.composite_primary_key? && association_primary_key.length != foreign_key.length
```
though I could be missing something